ext/drx/scatter_gather_aarch64.c

```diff
@@ -38,6 +38,479 @@
+         opnd_create_reg_element_vector(sg_info->index_reg, sg_info->element_size));
+
+    if (sg_info->is_load) {
+        /* ldr[bh]  scratch_gpr, [base_reg, scratch_gpr, mod #amount] */
+        opnd_t mem = opnd_create_base_disp_shift_aarch64(
+            sg_info->base_reg, scratch_gpr, sg_info->extend, sg_info->scaled, 0, 0,
+            sg_info->scalar_value_size, sg_info->extend_amount);
+
+        if (sg_info->is_scalar_value_signed) {
+            const reg_id_t ld_dst =
+                reg_resize_to_opsz(scratch_gpr, sg_info->element_size);
+            switch (sg_info->scalar_value_size) {
+            case OPSZ_1: EMIT(ldrsb, opnd_create_reg(ld_dst), mem); break;
+            case OPSZ_2: EMIT(ldrsh, opnd_create_reg(ld_dst), mem); break;
+            case OPSZ_4: EMIT(ldrsw, opnd_create_reg(ld_dst), mem); break;
+            default: DR_ASSERT_MSG(false, "Invalid scatter_gather_info_t data");
+            }
+        } else {
+            const reg_id_t scratch_gpr_w = reg_resize_to_opsz(scratch_gpr, OPSZ_4);
+            switch (sg_info->scalar_value_size) {
+            case OPSZ_1: EMIT(ldrb, opnd_create_reg(scratch_gpr_w), mem); break;
+            case OPSZ_2: EMIT(ldrh, opnd_create_reg(scratch_gpr_w), mem); break;
+            case OPSZ_4: EMIT(ldr, opnd_create_reg(scratch_gpr_w), mem); break;
+            case OPSZ_8: EMIT(ldr, opnd_create_reg(scratch_gpr), mem); break;
+            default: DR_ASSERT_MSG(false, "Invalid scatter_gather_info_t data");
+            }
+        }
+
+        /* cpy      gather_dst_reg.elsz, scratch_pred/m, scratch_gpr */
+        EMIT(cpy_sve_pred,
+             opnd_create_reg_element_vector(sg_info->gather_dst_reg,
+                                            sg_info->element_size),
+             opnd_create_predicate_reg(scratch_pred, true),
+             opnd_create_reg(reg_resize_to_opsz(scratch_gpr, sg_info->element_size)));
+    } else {
+        DR_ASSERT_MSG(sg_info->is_load, "Stores are not yet supported");
+    }
+
+    /* b        loop */
+    EMIT(b, opnd_create_instr(loop_label));
+
+    instrlist_meta_preinsert(bb, sg_instr, end_label);
+}
+
+/* Spill a scratch predicate or vector register.
+ * TODO i#3844: drreg does not support spilling predicate regs yet, so we do it
+ * ourselves.
+ * When that support is available, this funcion can be replaced with a drreg API call.
+ */
+reg_id_t
+reserve_sve_register(void *drcontext, instrlist_t *bb, instr_t *where,
+                     reg_id_t scratch_gpr, reg_id_t min_register, reg_id_t max_register,
+                     size_t slot_offset, opnd_size_t reg_size)
+{
+    /* Search the instruction for an unused register we will use as a temp. */
+    reg_id_t reg;
+    for (reg = min_register; reg <= max_register; ++reg) {
+        if (!instr_uses_reg(where, reg))
+            break;
+    }
+
+    drmgr_insert_read_tls_field(drcontext, tls_idx, bb, where, scratch_gpr);
```

Comment:
You could get directly-addressable spill space: this is an extra load.  But given the heavyweight unrolling here in general it is probably ok and some platforms do have limited direct slots.

Reply:
I did it this way mainly because that's what the x86 implementation does so it might not be the best solution for AArch64. I'm going to leave it as it is for now but I'll come back to it and look at optimising it as I build out support for the rest of the scatter/gather instructions.